Fix bug when an index appears in two expressions

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #105055
* #102611
* #104987
* #104975
* __->__ #104886

We were not adding the bounds to `replacement_vars` in this case.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8

lezcano committed:
Fix bug when an index appears in two expressions

We were not adding the bounds to `replacement_vars` in this case.

[ghstack-poisoned]

lezcano committed:
Update on "Fix bug when an index appears in two expressions"


We were not adding the bounds to `replacement_vars` in this case.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

lezcano committed:
Update on "Fix bug when an index appears in two expressions"


We were not adding the bounds to `replacement_vars` in this case.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

lezcano committed:
Update on "Fix bug when an index appears in two expressions"


We were not adding the bounds to `replacement_vars` in this case.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

lezcano committed:
Update on "Fix bug when an index appears in two expressions"


We were not adding the bounds to `replacement_vars` in this case.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

ezyang commented on torch/utils/_sympy/value_ranges.py:
```diff
@@ -83,6 +83,19 @@ def __contains__(self, x):
         x = simple_sympify(x)
         return sympy_generic_le(self.lower, x) and sympy_generic_le(x, self.upper)
 
+    def tighten(self, other):
```
What exactly does tighten do?

lezcano replied:
i.e. given two bounds, it returns the AND of them.

eellison replied:
add docstring / type annotation ?

eellison commented on torch/_inductor/bounds.py:
```diff
@@ -87,9 +87,10 @@ def set_indirect(self, old, new):
-            return bound
-
-        bound = bound_sympy(expr, self.replacement_vals)
+        if bound is None:
```
when would we have None bound as opposed to `ValueRanges.unknown()`

lezcano replied:
we're doing a `get()`, so when the key is not in the dictionary yet.

lezcano committed:
Update on "Fix bug when an index appears in two expressions"


We were not adding the bounds to `replacement_vars` in this case.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

Skylion007 commented on torch/_inductor/bounds.py:
```diff
@@ -87,9 +87,10 @@ def set_indirect(self, old, new):
-            return bound
-
-        bound = bound_sympy(expr, self.replacement_vals)
+        if bound is None:
+            bound = bound_sympy(expr, self.replacement_vals)
+        # The following assertion is true at the time of this writing
+        # We don't assert is as to not execute bound_sympy when bound is not None
+        # assert bound is None or bound == bound_sympy(expr, self.replacement_vals)
         self.replacement_vals[name] = bound
         return bound
```
Couldn't most of this method just be reduced to this one liner using set_default https://www.w3schools.com/python/ref_dictionary_setdefault.asp?
```suggestion
        return self.replacement_vals.setdefault(name, bound_sympy(expr, self.replacement_vals))
```

Skylion007 replied:
Actually, nvm that would always evaluate bound_sympy unfortunately which may be undesirable...

lezcano replied:
That would eagerly execute `bound_sympy`, and that's exactly what we're trying to avoid.

lezcano committed:
Update on "Fix bug when an index appears in two expressions"


We were not adding the bounds to `replacement_vars` in this case.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

lezcano committed:
Update on "Fix bug when an index appears in two expressions"


We were not adding the bounds to `replacement_vars` in this case.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

lezcano committed:
Update on "Fix bug when an index appears in two expressions"


We were not adding the bounds to `replacement_vars` in this case.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

lezcano commented:
@pytorchbot merge

lezcano commented:
@pytorchbot merge
